test/test_torch.py

```diff
@@ -4548,38 +4548,63 @@ def test_index_fill(self, device, dtype):
         self.assertEqual(0, x.index_fill(0, index, -1).dim())
         self.assertEqual(0, x.index_fill_(0, index, -1).dim())
 
-    def test_index_select(self, device):
-        for dtype in [torch.int, torch.long]:
-            src = torch.randn(3, 4, 5, device=device)
-            # Index can be duplicated.
-            idx = torch.tensor([2, 1, 0, 1, 2], dtype=dtype, device=device)
-            dest = torch.index_select(src, 0, idx)
-            self.assertEqual(dest.shape, (5, 4, 5))
-            for i in range(idx.size(0)):
-                self.assertEqual(dest[i], src[idx[i]])
-
-            # Check that 'out' is used correctly.
-            out = torch.randn(5 * 4 * 5, device=device)
-            dest = torch.index_select(src, 0, idx, out=out.view(5, 4, 5))
-            self.assertEqual(dest.shape, (5, 4, 5))
-            for i in range(idx.size(0)):
-                self.assertEqual(dest[i], src[idx[i]])
-            out.fill_(0.123)
-            self.assertEqual(out, dest.view(-1))  # Must point to the same storage.
-
-            # Bool tensor
-            src = torch.tensor([False, True, False, False], device=device, dtype=torch.bool)
-            idx = torch.tensor([1], dtype=dtype, device=device)
-            dest = torch.index_select(src, 0, idx)
-            self.assertEqual(torch.tensor([True]), dest)
-
-            # Complex Tensor
-            src = torch.randn(3, 4, 5, dtype=torch.complex64, device=device)
-            idx = torch.tensor([2, 1, 0, 1, 2], dtype=dtype, device=device)
-            dest = torch.index_select(src, 0, idx)
-            self.assertEqual(dest.shape, (5, 4, 5))
-            for i in range(idx.size(0)):
-                self.assertEqual(dest[i], src[idx[i]])
+    @dtypes(*torch.testing.get_all_dtypes())
+    def test_index_select(self, device, dtype):
+        num_src, num_dest = 3, 5
+
+        def make_arg(batch_sizes, n, dim, contig):
+            size_arg = batch_sizes[:dim] + (n,) + batch_sizes[dim:]
+            return make_tensor(size_arg, device, dtype, low=None, high=None, discontiguous=not contig)
+
+        def ref_index_select(src, dim, idx):
```

Comment:
Can this use np.take() as a reference?

Reply:
What is the reason for doing this? That seems extremely laborious to me, as `take` returns one element per index, while here we return one slice per indes. Or am I missing something here?

Reply:
`take` returns a slice if given an `axis` argument:
```
In [18]: torch.allclose(torch.from_numpy(np.take(a.numpy(), [1,0], axis=0)), a.index_select(0, torch.tensor([1,0])))                                                                                      
Out[18]: True

In [19]: torch.allclose(torch.from_numpy(np.take(a.numpy(), [1,0], axis=1)), a.index_select(1, torch.tensor([1,0])))                                                                                      
Out[19]: True

In [20]: torch.allclose(torch.from_numpy(np.take(a.numpy(), [1,0], axis=2)), a.index_select(2, torch.tensor([1,0])))                                                                                      
Out[20]: True
```

Reply:
Thank you!!